tensilelite/Tensile/SolutionStructs.py

```diff
@@ -1756,6 +1763,168 @@ def checkAndAssignWaveSeparateGlobalRead(state, tc):
+      doable = False
+    return doable
+
+  ########################################
+  # determine if current datatype can support DirectToVgpr
+  @staticmethod
+  def isDirectToVgprSupportDataType(state):
+    # So far, f32, f16 and bf16 only. TODO: enable DTV for other data types
+    #return (state["ProblemType"]["DataType"].isSingle() or state["ProblemType"]["DataType"].isDouble() or state["ProblemType"]["DataType"].isComplex() or \
+    #        state["ProblemType"]["DataType"].isHalf() or state["ProblemType"]["DataType"].isBFloat16() or state["ProblemType"]["DataType"].isInt8()) or \
+    #        state["ProblemType"]["DataType"].is8bitFloat()
+    return state["ProblemType"]["DataType"].isSingle() or state["ProblemType"]["DataType"].isHalf() or state["ProblemType"]["DataType"].isBFloat16()
+
+  ########################################
+  # determine can we use DirectToVgpr
+  @staticmethod
+  def isDirectToVgprDoable(state, tc):
+    MIindex = 0 if tc == 'A' else 1
+    numBytes = state["ProblemType"]["DataType"].numBytes()
+    # With MatrixInstruction only
+    if not state["EnableMatrixInstruction"] :
+      reject(state, "DirectToVgpr is for MatrixInstruction only")
+      return False
+
+    # disable the following combinations for initial implementation
+    # TODO: enable them
+    if state["LocalSplitU"] != 1:
+      reject(state, "LSU should be 1 for DirectToVgpr (tentative)")
+      return False
+    if state["MatrixInstB"] != 1:
+      reject(state, "MatrixInstBN should be 1 for DirectToVgpr (tentative)")
+      return False
+    if state["DirectToVgprA"] and state["DirectToVgprB"]:
+      reject(state, "DirectToVgprA + DirectToVgprB is not supported (tentative)")
+      return False
+    if state["StreamK"]:
+      reject(state, "DirectToVgpr does not work with StreamK (tentative)")
+      return False
+
+    # check if the DataType can support DirectToVgpr
+    if not Solution.isDirectToVgprSupportDataType(state):
+      reject(state, "no DirectToVgpr support for this input data type")
+      return False
+
+    # Does not work with TLU = False and PrefetchLocalRead = 0
+    if (not state["ProblemType"]["TLU%c"%tc]) and state["PrefetchLocalRead"] == 0:
+      reject(state, "DirectToVgpr%c does not supports TLU%c = False and PrefetchLocalRead = 0"%(tc, tc))
+      return False
+
+    # Does not work with TLU = False and CGEMM/DGEMM/DGEMM (not supported)
+    if (not state["ProblemType"]["TLU%c"%tc]) and (state["ProblemType"]["DataType"].isDouble() or \
+        state["ProblemType"]["DataType"].isComplex()):
+      reject(state, "DirectToVgpr%c does not supports TLU%c = False + S/C/D/ZGEMM"%(tc, tc))
+      return False
+
+    # numBytes < 4 case
+    if numBytes < 4:
+      # Does not work with TLU = True and numBytes < 4 (not supported)
+      if state["ProblemType"]["TLU%c"%tc]:
+        doable = Solution.isVgprForLocalReadPackingDoable(state)
```

Comment:
why do we need isVgprForLocalReadPackingDoable(state)? It is rejected always when TLU = True and numBytes < 4

Reply:
I removed the comment at line 1841 to avoid confusion.
  "# Does not work with TLU = True and numBytes < 4 (not supported)"

Reply:
This comment was from old implementation (before implementing DTV+packing) and no longer valid.